tensorflow/compiler/mlir/tosa/transforms/legalize_utils.cc

```diff
@@ -549,21 +549,50 @@ LogicalResult ApplyPatternsWithShapeResolution(
     return failure();
   }
 
+  IRRewriter rewriter(func.getContext());
+
   // Check that constant attributes types and op types match up. If the lowering
   // needs to change a type (e.g. fp16 -> fp32) its possible the return type
   // could be incorrect.
-  //
-  // This should be investigate for whether it is still necessary due to quant
-  // type stripping changing.
-  func.walk([&](tosa::ConstOp op) {
-    auto ety = op.value().getType().getElementType();
-    auto new_ty = op.getType().cast<ShapedType>().clone(ety);
-    op.getResult().setType(new_ty);
+  // In the case of quantization it checks the const result type and the attribute values
+  // type and makes sure that if one of them is quantized the quantization information
+  // is preserved and that it is propagated to where it is missing.
+  func.walk([&](tosa::ConstOp tosa_const_op) {
+    mlir::DenseElementsAttr value = tosa_const_op.value().cast<DenseElementsAttr>();
+
+    auto op_value_element_type  = value.getType().getElementType();
+    auto op_result_element_type = tosa_const_op.getResult().getType().cast<ShapedType>().getElementType();
+
+    bool value_quantized  = op_value_element_type.isa<quant::QuantizedType>();
+    bool result_quantized = op_result_element_type.isa<quant::QuantizedType>();
+
+    mlir::ShapedType new_return_type;
```

Comment:
These do not need to be declared here.

Reply:
I need these to check what is quantized so I need them here.

Reply:
Done